In [2]:
#installing pyspark
!pip install pyspark

     |████████████████████████████████| 281.4 MB 38 kB/s 
     |████████████████████████████████| 198 kB 60.9 MB/s 
  Created wheel for pyspark: filename=pyspark-3.2.1-py2.py3-none-any.whl size=281853642 sha256=4961119d78aaefae30daae7846f477a1894b2710c429e06b2b448ac8648ff613
  Stored in directory: /root/.cache/pip/wheels/9f/f5/07/7cd8017084dce4e93e84e92efd1e1d5334db05f2e83bcef74f
Successfully built pyspark


In [3]:
#importing libaries
import itertools
import pyspark

In [4]:
#making spark context instance
sc=pyspark.SparkContext("local","Apriori")
#loading data
data = sc.textFile("browsing.txt")

In [8]:
#reading data line by line
txt_RDD = data.map(lambda line:line.split())
txt_RDD.take(10)

[['FRO11987', 'ELE17451', 'ELE89019', 'SNA90258', 'GRO99222'],
 ['GRO99222',
  'GRO12298',
  'FRO12685',
  'ELE91550',
  'SNA11465',
  'ELE26917',
  'ELE52966',
  'FRO90334',
  'SNA30755',
  'ELE17451',
  'FRO84225',
  'SNA80192'],
 ['ELE17451', 'GRO73461', 'DAI22896', 'SNA99873', 'FRO86643'],
 ['ELE17451', 'ELE37798', 'FRO86643', 'GRO56989', 'ELE23393', 'SNA11465'],
 ['ELE17451',
  'SNA69641',
  'FRO86643',
  'FRO78087',
  'SNA11465',
  'GRO39357',
  'ELE28573',
  'ELE11375',
  'DAI54444'],
 ['ELE17451',
  'GRO73461',
  'DAI22896',
  'SNA99873',
  'FRO18919',
  'DAI50921',
  'SNA80192',
  'GRO75578'],
 ['ELE17451',
  'ELE59935',
  'FRO18919',
  'ELE23393',
  'SNA80192',
  'SNA85662',
  'SNA91554',
  'DAI22177'],
 ['ELE17451',
  'SNA69641',
  'FRO18919',
  'SNA90258',
  'ELE28573',
  'ELE11375',
  'DAI14125',
  'FRO78087'],
 ['ELE17451',
  'GRO73461',
  'DAI22896',
  'SNA80192',
  'SNA85662',
  'SNA90258',
  'DAI46755',
  'FRO81176',
  'ELE66810',
  'DAI49199',
  'DAI91535',
  'GRO9475

In [18]:
#taking counts of each id
from operator import add
individual_counts = data.flatMap(lambda line: line.split()).map(lambda x: (x,1)).reduceByKey(add)
individual_counts.take(10)

#getting frequency of the id
frequent_items_counts = individual_counts.filter(lambda x: x[1]>=100).sortBy(lambda x: x[0])
frequent_items_counts.take(10)

[('DAI11223', 155),
 ('DAI11778', 117),
 ('DAI13194', 116),
 ('DAI13266', 181),
 ('DAI13788', 213),
 ('DAI13902', 233),
 ('DAI14125', 163),
 ('DAI14181', 110),
 ('DAI14470', 153),
 ('DAI15681', 132)]

In [11]:
#most frequent singles(ID)
frequent_items_counts_collect = individual_counts.filter(lambda x: x[1]>=100).sortBy(lambda x: x[0]).collect()
frequent_items = frequent_items_counts.map(lambda x: x[0])
frequent_items_list = frequent_items_counts.map(lambda x: x[0]).collect()
frequent_items_list[:10]

['DAI11223',
 'DAI11778',
 'DAI13194',
 'DAI13266',
 'DAI13788',
 'DAI13902',
 'DAI14125',
 'DAI14181',
 'DAI14470',
 'DAI15681']

In [12]:
#defining a filter find out the data for pairs
def filtering_pair(l):
    ret=[]
    pair_lst=[]
    for i in l:
        if i in frequent_items_list:
            ret.append(i)
    ret.sort()
    for pair in itertools.combinations(ret,2):
        pair_lst.append((pair,1))
    return(pair_lst)

In [13]:
#defining confidence for pairs
def confidence_1(par):
    lst=[]
    conf1 = par[1]/(frequent_pair_counts.filter(lambda x: x[0]==par[0][0]).collect())[0][1]
    conf2 = par[1]/(frequent_items_counts.filter(lambda x: x[0]==par[0][1]).collect())[0][1]
    lst.append(((par[0],par[1]),conf1))
    lst.append(((par[1],par[0]),conf2))
    return(lst)

In [14]:
#defining a filter find out data for triplets
def filter_triplets(l):
    ret=[]
    triplet=[]
    for i in l:
        if i in frequent_items_2:
            ret.append(i)
    ret.sort()
    for trip in itertools.combinations(ret,3):
        triplet.append(((trip),1))
    return(triplet)

In [15]:
#defining confidence for triplets
def confidence_2(trip):
    confidence_for_2=[]
    for i in (itertools.combinations(trip[0],2)):
        conf=trip[1]/(frequent_pair_counts.filter(lambda X: x[0]==i)).collect[0][1]
        confidence_for_2.append((i,conf))
    return(confidence_for_2)

In [16]:
#most frequent doubles
#calculting pairs of each id 
pair_count = txt_RDD.flatMap(lambda x: filtering_pair(x)).reduceByKey(add).sortBy( lambda x: x[0])
pair_count.take(10)

#getting ferquent ppair count
frequent_pair_counts = pair_count.filter(lambda x: x[1]>=100)
frequent_pair_counts.take(10)
frequent_pair_counts_list = frequent_pair_counts.collect()


#pairs bought together
frequent_pair = frequent_pair_counts.map(lambda x: x[0])
# frequent_pair.take(10)
frequent_pair_counts_list[:10]

[(('DAI16732', 'FRO78087'), 106),
 (('DAI18527', 'SNA44451'), 102),
 (('DAI22177', 'DAI31081'), 127),
 (('DAI22177', 'DAI62779'), 382),
 (('DAI22177', 'DAI63921'), 136),
 (('DAI22177', 'DAI75645'), 123),
 (('DAI22177', 'DAI83733'), 126),
 (('DAI22177', 'DAI85309'), 172),
 (('DAI22177', 'ELE17451'), 203),
 (('DAI22177', 'ELE26917'), 134)]

In [17]:
#most frequent triplets
#list of most frequent items for triplets
frequent_items_2 = frequent_pair.flatMap(lambda x: x).collect()
frequent_items_2[:10]

#calculating items counts for triplets
triplets_count = txt_RDD.flatMap(lambda x: filter_triplets(x)).reduceByKey(add).sortBy(lambda x: x[0])
triplets_count.take(10)

#calculating frequent seen triplets
frequent_triplets = triplets_count.filter(lambda x: x[1]>=100)
frequent_triplets.take(10)

[(('DAI22896', 'DAI62779', 'GRO73461'), 101),
 (('DAI23334', 'DAI62779', 'ELE92920'), 143),
 (('DAI31081', 'DAI62779', 'ELE17451'), 103),
 (('DAI31081', 'DAI75645', 'FRO40251'), 122),
 (('DAI31081', 'ELE32164', 'GRO59710'), 112),
 (('DAI31081', 'FRO40251', 'GRO85051'), 102),
 (('DAI31081', 'FRO40251', 'SNA80324'), 103),
 (('DAI42083', 'DAI62779', 'DAI92600'), 105),
 (('DAI42083', 'DAI92600', 'ELE17451'), 117),
 (('DAI42493', 'DAI62779', 'ELE17451'), 112)]

In [24]:
#calculating confidence for pairs
confidence_one=[]
for i in frequent_pair_counts_list:
    confid_1 = i[1]/(frequent_items_counts.filter(lambda x: x[0]==i[0][0]).collect())[0][1]
    confid_2 = i[1]/(frequent_items_counts.filter(lambda x: x[0]==i[0][1]).collect())[0][1]
    confidence_one.append(((i[0][0],i[0][1]),confid_1))
    confidence_one.append(((i[0][1],i[0][0]),confid_2))
#top 5 values for pairs
confidence_one.sort(key=lambda x: -x[1])
confidence_one[:5]    

[(('DAI93865', 'FRO40251'), 1.0),
 (('GRO85051', 'FRO40251'), 0.999176276771005),
 (('GRO38636', 'FRO40251'), 0.9906542056074766),
 (('ELE12951', 'FRO40251'), 0.9905660377358491),
 (('DAI88079', 'FRO40251'), 0.9867256637168141)]

'DAI93865', 'FRO40251' -> 1.0

'GRO85051', 'FRO40251' -> 0.999176276771005

'GRO38636', 'FRO40251' -> 0.9906542056074766

'ELE12951', 'FRO40251' -> 0.9905660377358491

'DAI88079', 'FRO40251' -> 0.9867256637168141

In [31]:
#calculating confidence for triplets
frequent_triplets_list = triplets_count.filter(lambda x: x[1]>=100).collect()
confidence_triplets=[]
for i in frequent_triplets_list:
    a=set(i[0])
    for j in (itertools.combinations(i[0],2)):
        b=set(j)
        c=list(a-b)
        confidence_=i[1]/(frequent_pair_counts.filter(lambda x: x[0]==j)).collect()[0][1]
        confidence_triplets.append(((j,c[0]),confidence_))

In [33]:
#top 5 values for triplets
confidence_triplets.sort(key = lambda x: -x[1])
confidence_triplets[:5]

[((('DAI23334', 'ELE92920'), 'DAI62779'), 1.0),
 ((('DAI31081', 'GRO85051'), 'FRO40251'), 1.0),
 ((('DAI55911', 'GRO85051'), 'FRO40251'), 1.0),
 ((('DAI62779', 'DAI88079'), 'FRO40251'), 1.0),
 ((('DAI75645', 'GRO85051'), 'FRO40251'), 1.0)]

'DAI23334', 'ELE92920', 'DAI62779' -> 1.0

'DAI31081', 'GRO85051', 'FRO40251' -> 1.0

'DAI55911', 'GRO85051', 'FRO40251' -> 1.0

'DAI62779', 'DAI88079', 'FRO40251' -> 1.0

'DAI75645', 'GRO85051', 'FRO40251' -> 1.0